### PR TITLE
Pass group ranks and options to third party distributed backends

### LINKED ISSUE #73163: Support Passing Participating Group Ranks and Options to 3rd Party Distributed Backends
### 🚀 The feature, motivation and pitch

While working on a 3rd party distributed backend, I noticed PyTorch's [`_new_process_group_helper()`](https://github.com/pytorch/pytorch/blob/9f541aa3aca768e7fbfa4a9d648b554f22b261f7/torch/distributed/distributed_c10d.py#L633) does not pass group's participating ranks to the backend. 

Also, it does not support passing any process group options as well. This is confirmed as a missing feature in the discussion forum https://discuss.pytorch.org/t/getting-subgroup-ranks-info-in-custom-backend/128067

### Alternatives

_No response_

### Additional context

_No response_

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### PR DESCRIPTION
**Pull request by esaliya**
Fixes #73163

PyTorch's [_new_process_group_helper()](https://github.com/pytorch/pytorch/blob/9f541aa3aca768e7fbfa4a9d648b554f22b261f7/torch/distributed/distributed_c10d.py#L633) does not pass group's participating ranks to the backend.

This PR adds the above capability. Also, refactors some variables for better clarity.

### CODE REVIEW

File: torch/distributed/distributed_c10d.py
Comment (pritamdamania87) [inline torch/distributed/distributed_c10d.py:658]: What is the reason to remove the group_count increment from here and move it below?

File: torch/distributed/distributed_c10d.py
Comment (pritamdamania87) [inline torch/distributed/distributed_c10d.py:777]: I guess adding `global_ranks_in_group` and `pg_options` might be fine here, but it doesn't seem like passing in `_group_count` is a good idea. This is an internal implementation details that we should not expose.

There also might be a BC issue here where we would break existing `backend_creator_fns` with this change.

File: torch/distributed/distributed_c10d.py
Comment (pritamdamania87) [inline torch/distributed/distributed_c10d.py:774]: Why do we need all of this? Can't we just pass in `None` to `backend_creator_fn` instead?

**Review by pritamdamania87 — COMMENTED** [review]
@esaliya It looks like there some CI failures.

cc @mrshenli Since you are more familiar with c10d extension.

File: torch/distributed/distributed_c10d.py
Comment (esaliya) [reply torch/distributed/distributed_c10d.py:658]: Here the `_group_count` is updated only if the group does not have a name. The change is made to use `_group_count` to represent the total number of groups created. This has two cases. The first is when a non-participating rank tries to create a group. In this case, the `_group_count` is immediately incremented and the `GroupMember.NON_GROUP_MEMBER` is returned.

The second is when a legitimate process group is created. In this case, the `_group_count`` is incremented after the group is created at the end.

File: torch/distributed/distributed_c10d.py
Comment (esaliya) [reply torch/distributed/distributed_c10d.py:774]: The group options have to be a custom class extending from the `ProcessGroup::Options`.  In the case of 3rd party extensions, this requires adding a pybind definition like the following. 

```
py::class_<c10d::ProcessGroupTest::Options, c10::intrusive_ptr<c10d::ProcessGroupTest::Options>>(m, "Options")
            .def(py::init<const bool &>())
            .def("create", &c10d::ProcessGroupTest::Options::create)
            .def_readwrite("some_test_property", &c10d::ProcessGroupTest::Options::some_test_property);
```

Passing None in this cause pybind to fail not being able to create an instance of the custom options type

File: torch/distributed/distributed_c10d.py
Comment (esaliya) [reply torch/distributed/distributed_c10d.py:777]: The `_group_count` is used as a group identifier. When a collective is called on a specific process group object, for 3rd party backends there's currently no support to associate from what group is the call coming from. Including a group identifier bridges this gap.

Agree with BC issue, and happy to hear suggestions here. `global_ranks_in_group` and `pg_options` are currently being not passed to 3rd party, which is the main limitation trying to address in this PR.

**Comment by esaliya** [comment]
> @esaliya It looks like there some CI failures.
Seems there are some test backends that fail due to the new API. I'll take a look to fix this.

File: torch/distributed/distributed_c10d.py
Comment (esaliya) [reply torch/distributed/distributed_c10d.py:777]: I want to check back on this point on the point about backward compatibility here. We can incorporate an API version number in 3rd party distributed backends and based on that invoke the existing signature or the new one proposed. Does that sound good?

File: torch/distributed/distributed_c10d.py
Comment (kumpera) [reply torch/distributed/distributed_c10d.py:774]: We should leave for the backend creation function the to fill in with a default options object.

This eliminates this BC breaking requirement that there must be a class named `Option` in the same module as the creation function.

File: torch/distributed/distributed_c10d.py
Comment (kumpera) [reply torch/distributed/distributed_c10d.py:777]: We should reorder the arguments so the new ones are at the end. 
Instead of breaking existing backends, we should check the number of arguments the creation function takes - and emit a warning if too few.

**Review by kumpera — CHANGES_REQUESTED** [review]

File: torch/distributed/distributed_c10d.py
Comment (esaliya) [reply torch/distributed/distributed_c10d.py:774]: This has been fixed and will be pushed in the next iteration

File: torch/distributed/distributed_c10d.py
Comment (esaliya) [reply torch/distributed/distributed_c10d.py:777]: @kumpera Alternatively, does extending `dist.register_backend()` to take in an API version (string) info sound reasonable to support backward compatibility? This would be an optional argument, so during dispatching if version is omitted or say v1.0 we can send only the limited number of arguments as done today.

Comment (esaliya) [reply torch/distributed/distributed_c10d.py:777]: I want to get some thoughts to unblock this PR. The main issue is the inability to inspect the function signature at runtime for the backend creator function. Python inspect module fails for the same reason as in this open issue [28233](https://github.com/pytorch/pytorch/issues/28233). Also, the definition and python binding of this function are in the control of the 3rd party provider.

So in this regard, what about specifically asking the 3rd party provider to specify the API version. For existing backends we can assume the API to be v1.0, which accepts the limited set of arguments.

If this sounds good, I can provide the implementation in the next iteration. Let me know.

Comment (kumpera) [reply torch/distributed/distributed_c10d.py:777]: That sounds the path forward. Do you think something along the lines of:
```
class BackendProtocol(Enum):
 ONE = auto() # original 4 args func
 TWO = auto() # new version with 6 args
class Backend:
    def register_backend(cls, name, func, version=BackendProtocol.One):
      ....
```

Comment (esaliya) [reply torch/distributed/distributed_c10d.py:777]: Yes, this is along the lines of what I meant. 
I also have another approach that might be little extensible than this. That would be to define an overload `register_backend` function that'll accept a function that takes a predefined struct instead of individual arguments. The struct can be defined like how ProcessGroupOptions or Store classes are defined. 

I'll make an update, so you can take a pass.

Comment (esaliya) [reply torch/distributed/distributed_c10d.py]: I've added a flag, `extended_api` to differentiate between old and new APIs when registering 3P communication backends. When using the `extended_api=True` the registered backend is expected to accept a new `c10d::DistributedBackendOptions` instance along with any process group options implemented by that backend extending the `c10d::ProcessGroup::Options`. 

The `c10d::DistributedBackendOptions` gives the ability to add future changes to the backend registration contract without breaking backward compatibility.

**Comment by esaliya** [comment]
@kumpera could you help to trigger the CI for this PR?

File: test/allowlist_for_publicAPI.json
Comment (albanD) [inline test/allowlist_for_publicAPI.json:378]: You should not add new entries to this file in general.
You will need to properly only expose the public functions and make non-public ones properly private as per the recommendation from the failing test.

**Review by albanD — CHANGES_REQUESTED** [review]

File: test/allowlist_for_publicAPI.json
Comment (esaliya) [reply test/allowlist_for_publicAPI.json:378]: This is removed for the next iteration

**Comment by esaliya** [comment]
@JackCaoG, got a suggestion you might have some tip here. I am making a change to 3rd party backend registration in `torch.distributed.register_backend()`. This cause test failures coming from xla repo. What's the recommended way to fix this? I know the change to be done in xla test but that would need the changes coming from this PR (a bit like a chicken-and-egg here). Any thoughts?

**Comment by alanwaketan** [comment]
> @JackCaoG, got a suggestion you might have some tip here. I am making a change to 3rd party backend registration in `torch.distributed.register_backend()`. This cause test failures coming from xla repo. What's the recommended way to fix this? I know the change to be done in xla test but that would need the changes coming from this PR (a bit like a chicken-and-egg here). Any thoughts?

I can take a look and investigate. Can you ping me again when the upstream has approved the change?

**Comment by esaliya** [comment]
> > @JackCaoG, got a suggestion you might have some tip here. I am making a change to 3rd party backend registration in `torch.distributed.register_backend()`. This cause test failures coming from xla repo. What's the recommended way to fix this? I know the change to be done in xla test but that would need the changes coming from this PR (a bit like a chicken-and-egg here). Any thoughts?
> 
> I can take a look and investigate. Can you ping me again when the upstream has approved the change?

@alanwaketan did you mean when this PR is merged -- it'll be blocked on these test failures?

**Comment by JackCaoG** [comment]
I think the idea is to simplify the pytorch/xla fix. Often time pytorch/xla fix the CI issue with a companion pr but then the pytorch pr changed due to some review comments from pytorch reviewer. PyTorch/XLA will need to update its fix pr again.

I generally prefer you to ignore the pytorch/xla CI and fix other review comments and get approval from pytorch side reviewer. Pytorch/xla contributor can fix our CI failure after everything else is ready and final.

File: test/allowlist_for_publicAPI.json
Comment (kumpera) [inline test/allowlist_for_publicAPI.json:453]: We're transition c10d to be fully typed with https://github.com/pytorch/pytorch/pull/84119

File: torch/csrc/distributed/c10d/Types.hpp
Comment (kumpera) [inline torch/csrc/distributed/c10d/Types.hpp:77]: Group ID should be a string and use `group_name` from the python side. We don't want to leak the details that this is a counter as it's going to change in the near future to a random string.

File: torch/distributed/distributed_c10d.py
Comment (kumpera) [inline torch/distributed/distributed_c10d.py:689]: We should keep all usage of `_group_count` together and at the beginning of this function to ensure consistency and correct usage (it's going to be easy to miss an increment when another early exit is added).

**Review by kumpera — CHANGES_REQUESTED** [review]
Few minor changes needed, but the overall design looks good.
Thanks for taking the time on this change.

File: torch/distributed/distributed_c10d.py
Comment (esaliya) [reply torch/distributed/distributed_c10d.py]: Reverted `_group_count` changes and used `group_name` as group identifier for 3rd party backends.

File: torch/csrc/distributed/c10d/Types.hpp
Comment (esaliya) [reply torch/csrc/distributed/c10d/Types.hpp]: Done

File: torch/distributed/distributed_c10d.py
Comment (kumpera) [inline torch/distributed/distributed_c10d.py:871]:
```suggestion
            extended_api = backend_plugin.extended_api
```

File: torch/distributed/distributed_c10d.py
Comment (kumpera) [inline torch/distributed/distributed_c10d.py:876]: Is this a vestigial change? I don't see any other mention of Options anywhere else in this change.

**Review by kumpera — CHANGES_REQUESTED** [review]
Looks great, let's address those two minor issues and merge it.

**Review by kumpera — APPROVED** [review]
LGTM, let's get this merged.

**Comment by JackCaoG** [comment]
Does this change still breaks XLA?

**Comment by esaliya** [comment]
> 

Technically, it should. I am waiting for CI pipelines to go through to see it.

**Comment by esaliya** [comment]
@JackCaoG I created a companion PR in xla with the fix to its failing test cases at https://github.com/pytorch/xla/pull/3982. Let me know how to get this merged, so the PR here can go through.

**Comment by alanwaketan** [comment]
Hi @esaliya, can you patch your PR with hash: 2f1b8ffa18e1c376f51bb62deea0f545b4f9cbee on https://github.com/pytorch/pytorch/blob/73164/.github/ci_commit_pins/xla.txt#L1.

**Comment by alanwaketan** [comment]
It should correctly refer to https://github.com/pytorch/xla/pull/3983 and fix the CI. Once CI on both PR are green, you are good to land the PR on pytorch first. Thanks for the contribution.

**Comment by kumpera** [comment]
@albanD the public API issue is being sorted out by #84139

**Comment by esaliya** [comment]
> Hi @esaliya, can you patch your PR with hash: 2f1b8ffa18e1c376f51bb62deea0f545b4f9cbee on https://github.com/pytorch/pytorch/blob/73164/.github/ci_commit_pins/xla.txt#L1.

@alanwaketan I've updated this here in this PR now

File: .github/ci_commit_pins/xla.txt
Comment (kumpera) [inline .github/ci_commit_pins/xla.txt:1]: merge damage :/

Comment (esaliya) [reply .github/ci_commit_pins/xla.txt]: thanks for catching this. Fixed now

File: .github/ci_commit_pins/xla.txt
Comment (alanwaketan) [inline .github/ci_commit_pins/xla.txt:1]: I just reabse my PR at https://github.com/pytorch/xla/pull/3983.

And here is updated hash: 06c3224a7ecfaf268434fd56c2eb3f9bc973f223.

Comment (kumpera) [reply .github/ci_commit_pins/xla.txt]: thanks. updated the PR.

Comment (alanwaketan) [reply .github/ci_commit_pins/xla.txt]: I rebased again to see if we can fix the build error. Here is the new hash: 743d9fd4d2311fdd514caa977da2b7f29948258d.

Comment (esaliya) [reply .github/ci_commit_pins/xla.txt]: Thanks @alanwaketan, I updated the xla.txt in this PR

**Comment by alanwaketan** [comment]
There are so many companion checkins recently. Let me rebase again to see if that resolve the build failure.

File: .github/ci_commit_pins/xla.txt
Comment (alanwaketan) [inline .github/ci_commit_pins/xla.txt:1]: Here is the new one again. Hopefully we can merge the change this time.
96c96ce89d2731846b733054f3cef8915f31a609

Comment (esaliya) [reply .github/ci_commit_pins/xla.txt:1]: Thanks @alanwaketan

**Comment by alanwaketan** [comment]
Can you do a rebase as well? I think we need this master change: https://github.com/esaliya/pytorch/commit/a0b1693996f408c112b7a628860b7a754aaa4f77.

File: .github/ci_commit_pins/xla.txt
Comment (alanwaketan) [inline .github/ci_commit_pins/xla.txt:1]: This is the wrong commit hash. Please use the following.
96c96ce89d2731846b733054f3cef8915f31a609

**Comment by kumpera** [comment]
ONNX failure is unrelated.
test_noncontiguous_samples_nn_functional_conv_transpose2d_cuda_float32 failure is unrelated
functorch failure is unrelated

**Comment by alanwaketan** [comment]
Finally! Thanks Rodrigo and Saliya for keeping the XLA bots green.

**Pull request closed without merging**
